taint-mode: Stop unifying source-sink metavariables

IagoAbal opened the pull request:
At least it will not be the default. Forcing unification of source-sink
metavariables turned out to be confusing even for r2c's security team!
But we never had a real use for it, we just wanted to expose the
metavariables to the rule message, which we can achieve in other ways.
Here we chose to do the union of both metavariable sets, but prefer
the sink ones in case of collision.

And restore most of "taint-mode: Propagate metavariables correctly (#4877)"
We still have to propagate metavariables correctly!

Closes #4464

test plan:
make test # test included

PR checklist:

- [ ] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

IagoAbal commented:
Note that CI failure is expected until returntocorp/semgrep-rules#1941 is merged.

nmote commented on CHANGELOG.md:
```diff
@@ -41,6 +49,8 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
   constant will also be considered constant (#4301)
 - Constant propagation now recognizes operators `++` and `--` as side-effectful
   (#4667)
+- taint-mode: Metavariables bound by a `pattenr-inside` are now available to the
```
```suggestion
- taint-mode: Metavariables bound by a `pattern-inside` are now available to the
```

emjin commented on semgrep-core/src/engine/Dataflow_tainting.ml:
```diff
@@ -184,28 +185,49 @@ let report_findings env findings =
+  in
+  Some (source_mvars' @ sink_mvars)
+
+let merge_source_sink_mvars env =
```
The names of these functions are a little too compact to serve as descriptions. I think if I didn't already have context, even with the comments already here, I would be a bit confused about what's going on. Would be nice to have a function-level comment stating that we're merging the metavariables bound in the source pattern and the sink pattern.

Also, what happens to metavariables in sanitizers?

IagoAbal replied:
I believe we were also meant to unify source-sanitizer metavariables, but it seems that we never did. I added a `TODO` comment in a previous PR.

emjin commented on semgrep-core/src/engine/Match_tainting_rules.ml:
```diff
@@ -159,13 +160,9 @@ let taint_config_of_rule default_config equivs file ast_and_errors
                  || List.exists
-                      (fun rng' ->
-                        rng'.Range_with_metavars.r = rng.Range_with_metavars.r)
+                      (fun rng' -> rng'.RM.r = rng.RM.r)
```
I know you're just making a small change here, but I think using the alias for `Range_with_metavars` really makes this alphabet soup. Can we use a more descriptive name for `rng`?

IagoAbal replied:
I replaced it with `range`.

emjin reviewed (approved):
I kind of think it's going to be confusing either way, but we can try this. Maybe the real move is to forbid users from using the same metavariable in their sources and sinks unless they include an "I know what I'm doing I swear" option 😆

IagoAbal commented:
> I kind of think it's going to be confusing either way, but we can try this. Maybe the real move is to forbid users from using the same metavariable in their sources and sinks unless they include an "I know what I'm doing I swear" option 😆

If done from the beginning, I agree that it could have been a good approach! It's just a little bit inconvenient having to write `$FUNC1`, `$FUNC2`, and so on... but for sure it leaves no room for confusion.

Another option was to make it work similar-ish to join mode, and have a field to choose what variables to unify. You could have a qualifier to refer to each metavariable, as in `sources.$X`. After all, `pattern-sources` and `pattern-sinks` are two separate queries that are run independently. And this is not just an implementation detail!

I guess `taint_unify_mvars: true` somehow means "I know what I'm doing I swear"? 😄

IagoAbal committed:
taint-mode: Stop unifying source-sink metavariables

At least it will not be the default. Forcing unification of source-sink
metavariables turned out to be confusing even for r2c's security-research
team!

We never had a real use case for unification (although it seems there are
some), and we just wanted to expose the metavariables to the rule message,
which we can achieve in other ways. Here we chose to do the union of both
metavariable sets, but prefer the sink ones in case of collision.

And restore most of "taint-mode: Propagate metavariables correctly (#4877)"
We still have to propagate metavariables correctly!

Closes #4464 (again)

test plan:
make test # test included

IagoAbal committed:
Emma's review
